controller/src/main/java/io/pravega/controller/store/stream/AbstractInt96Counter.java

```diff
@@ -0,0 +1,127 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import io.pravega.common.lang.AtomicInt96;
+import io.pravega.common.lang.Int96;
+import io.pravega.common.tracing.TagLogger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+
+/**
+  * Abstract Int96 counter. It implements getNextCounter using the Int96Counter interface.
+  */
+abstract class AbstractInt96Counter implements Int96Counter {
+    private static final TagLogger log = new TagLogger(LoggerFactory.getLogger(AbstractInt96Counter.class));
+    final Object lock;
+    @GuardedBy("lock")
+    final AtomicInt96 limit;
+    @GuardedBy("lock")
+    final AtomicInt96 counter;
+    @GuardedBy("lock")
+    CompletableFuture<Void> refreshFutureRef;
+
+    public AbstractInt96Counter() {
+        this.lock = new Object();
+        this.counter = new AtomicInt96();
+        this.limit = new AtomicInt96();
+        this.refreshFutureRef = null;
+    }
+
+    @Override
+    public CompletableFuture<Int96> getNextCounter(OperationContext context) {
+        CompletableFuture<Int96> future;
+        synchronized (lock) {
+            Int96 next = counter.incrementAndGet();
+            if (next.compareTo(limit.get()) > 0) {
+                // ignore the counter value and after refreshing call getNextCounter
+                future = refreshRangeIfNeeded(context).thenCompose(x -> getNextCounter(context));
+            } else {
+                future = CompletableFuture.completedFuture(next);
+            }
+        }
+        return future;
+    }
+
+    @VisibleForTesting
+    CompletableFuture<Void> refreshRangeIfNeeded(final OperationContext context) {
+        CompletableFuture<Void> refreshFuture;
+        synchronized (lock) {
+            // Ensure that only one background refresh is happening. For this we will reference the future in refreshFutureRef
+            // If reference future ref is not null, we will return the reference to that future.
+            // It is set to null when refresh completes.
+            refreshFuture = this.refreshFutureRef;
+            if (this.refreshFutureRef == null) {
+                // no ongoing refresh, check if refresh is still needed
+                if (counter.get().compareTo(limit.get()) >= 0) {
+                    log.info("Refreshing counter range. Current counter is {}. Current limit is {}", counter.get(), limit.get());
+
+                    // Need to refresh counter and limit. Start a new refresh future. We are under lock so no other
+                    // concurrent thread can start the refresh future.
+                    refreshFutureRef = getRefreshFuture(context)
+                            .exceptionally(e -> {
+                                // if any exception is thrown here, we would want to reset refresh future so that it can be retried.
+                                synchronized (lock) {
```

Comment:
This is chaining futures under a lock. This could deadlock with itself under the right conditions.